src/sentry/integrations/bitbucket_server/repository.py

```diff
@@ -0,0 +1,113 @@
+        except Exception as e:
+            installation.raise_error(e)
+        else:
+            config["external_id"] = six.text_type(repo["id"])
+            config["name"] = repo["project"]["name"] + "/" + repo["name"]
+        return config
+
+    def build_repository_config(self, organization, data):
+        installation = self.get_installation(data.get("installation"), organization.id)
+        client = installation.get_client()
+        parts = data["identifier"].split("/")
+
+        try:
+            resp = client.create_hook(
+                parts[0],
+                parts[1],
+                {
+                    "name": "sentry-bitbucket-server-repo-hook",
+                    "url": absolute_uri(
+                        reverse(
+                            "sentry-extensions-bitbucketserver-webhook",
+                            kwargs={
+                                "organization_id": organization.id,
+                                "integration_id": data.get("installation").encode("utf-8"),
+                            },
+                        )
+                    ),
+                    "active": True,
+                    "events": ["repo:refs_changed", "pr:merged"],
+                },
+            )
+        except Exception as e:
+            installation.raise_error(e)
+        else:
+            return {
+                "name": data["identifier"],
+                "external_id": data["external_id"],
+                "url": installation.model.metadata["base_url"]
+                + "/projects/%s/repos/%s/browse" % (parts[0], parts[1]),
+                "config": {"name": data["identifier"], "webhook_id": resp["id"]},
+                "integration_id": data["installation"],
+            }
+
+    def on_delete_repository(self, repo):
+        installation = self.get_installation(repo.integration_id, repo.organization_id)
+        client = installation.get_client()
+
+        parts = repo.config["name"].split("/")
+
+        try:
+            client.delete_hook(parts[0], parts[1], repo.config["webhook_id"])
+        except ApiError as exc:
+            if exc.code == 404:
+                return
+            raise
+
+    def _format_commits(self, repo, commit_list):
+        return [
+            {
+                "id": c["id"],
+                "repository": repo.name,
+                "author_email": c["author"]["emailAddress"],
+                "author_name": c["author"]["name"],
+                "message": c["message"],
+                "timestamp": self.format_date(c["authorTimestamp"]),
+                "patch_set": None,
+            }
+            for c in commit_list
+        ]
+
+    def compare_commits(self, repo, start_sha, end_sha):
+        installation = self.get_installation(repo.integration_id, repo.organization_id)
+        client = installation.get_client()
+        # use config name because that is kept in sync via webhooks
+        parts = repo.config["name"].split("/")
+        try:
```

Comment:
@pierredup I see in BitBucket Cloud we have a check if `start_sha` is `None`. Wh don't we need the check here?

Reply:
> I see in BitBucket Cloud we have a check if start_sha is None. Why don't we need the check here?

`start_sha` won't ever be `None`, as the hook just sends the value as `0000000000000000000000000000000000000000` when it's empty.

> I noticed that when my first commit (but not subsequent commits) produced an error on the webhook:

This is definitely related, I'll have a look at this and see how we can handle an empty commit